### PR TITLE
Feat/new wheel picker integration

### PR DESCRIPTION
**Pull request by mendyEdri**
## Description
Integrate the new `WheelPicker` into our Picker when passing `useNativePicker` prop.

WIP:
- <s>Unwanted extra space in the `WheelPicker` when it's under the `safeArea`.</s>
- <s>FlatList's `initialScrollIndex` doesn't work, hacked with timer for now (could be a permanent fix)</s>.
- <s>WheelPicker is not fully controlled; when passing a hard-coded data it still change when scrolling the WheelPicker, 
example: `<WheelPicker selectedValue={‘Java’} />` when scrolling, it's not scrolling back to the 'Java' row.</s>
- <s>Implement and check behaviour on Android.</s>
- <s>Extract finding index to presenter/other type.</s>
- <s>add unit test for WheelPicker presenter.</s>

## Changelog
Integrate the new customizable `WheelPicker` into our Picker when passing `useNativePicker` prop.

### CODE REVIEW

File: src/components/picker/PickerDialog.ios.js
Comment (ethanshar) [inline src/components/picker/PickerDialog.ios.js:59]: @mendyEdri 
Regarding 
> Unwanted extra space in the WheelPicker when it's under the safeArea.

Consider doing this 

```
      <Dialog {...dialogProps} width="100%" migrate bottom animationConfig={{duration: 300}} panDirection={panDirection}>
        <View flex bg-white>
          {this.renderHeader()}
          <View centerV flex>
            {this.renderPicker()}
          </View>
          <View useSafeArea/>
        </View>
      </Dialog>
```

I added `<View useSafeArea/>` under the wrapper View of the picker.

**Comment by ethanshar** [comment]
> * Unwanted extra space in the `WheelPicker` when it's under the `safeArea`.

See my comment in the code

> * FlatList's `initialScrollIndex` doesn't work, hacked with timer for now (could be a permanent fix).

Instead of hacking with setTimeout. Try scrolling to initial index on FlatList's `onLayout` prop. 

> * WheelPicker is not fully controlled; when passing a hard-coded data it still change when scrolling the WheelPicker,
>   example: `<WheelPicker selectedValue={‘Java’} />` when scrolling, it's not scrolling back to the 'Java' row.

I don't think WheelPicker can ever be fully controlled because of the way users interact with it. 
We can't have the wheel's value to wait for prop change, the interaction is too fast - so it must be uncontrolled in that way. 
Having said that, the dialog picker can be controlled. cause the final value is determined only when pressing 'done'.

**Comment by mendyEdri** [comment]
> the dialog picker can be
@ethanshar 
Agree, user interaction can constantly change the `value`.
though, I tried to mimic the current behaviour of WheelPicker, that was re-scrolling to `value` if it was controlled, so is it ok that the behaviour will differ?

**Comment by ethanshar** [comment]
> > the dialog picker can be
> 
> Agree, user interaction can constantly change the `value`.
> though, I tried to mimic the current behaviour of WheelPicker, that was re-scrolling to `value` if it was controlled, so is it ok that the behaviour will differ?
> 2nd: when w'll have a few pickers, wi'll need a way to control the value.
> (ex: min to max, max can't be lower then min)

I see what you mean.. 
I believe you want to treat the value prop as an initialValue and if the user change it in runtime it will reset to that value, right?

**Comment by mendyEdri** [comment]
> and if the user change it in runtime it will reset to that value,

right. we probably going to need it

File: src/components/picker/NativePicker.js
Comment (ethanshar) [inline src/components/picker/NativePicker.js:71]: Not sure this is required.. 
usually Views are stretched horizontally (unless you change their flexDirection to be row)
So the width actually determined by the Dialog width

File: src/incubator/WheelPicker/index.tsx
Comment (ethanshar) [inline src/incubator/WheelPicker/index.tsx:14]: you "can't" import from `react-native-ui-lib` in source files. 
You can get `Spacings` from this import 
`import {Colors} from '../../../src/style';`

File: src/incubator/WheelPicker/index.tsx
Comment (ethanshar) [inline src/incubator/WheelPicker/index.tsx:42]: Don't you prefer to keep it synced with the Picker's `onChange` API?

File: src/incubator/WheelPicker/index.tsx
Comment (ethanshar) [inline src/incubator/WheelPicker/index.tsx:81]: Is `currentIndex` actually represents the initial index?

File: src/incubator/WheelPicker/index.tsx
Comment (ethanshar) [inline src/incubator/WheelPicker/index.tsx:148]: Is there any benefit from using `useMemo` here (for `fader`)? 
Why not just render `<Fader....` twice at the bottom

File: src/incubator/WheelPicker/usePresenter.tsx
Comment (ethanshar) [inline src/incubator/WheelPicker/usePresenter.tsx:50]: at any point when swiping the wheel picker and `selectedValue` stays the same this will return true

File: src/incubator/WheelPicker/index.tsx
Comment (ethanshar) [inline src/incubator/WheelPicker/index.tsx:99]: Verify that initial value works.. right now it always takes the first one.

File: src/incubator/WheelPicker/index.tsx
Comment (ethanshar) [inline src/incubator/WheelPicker/index.tsx:64]: use React.Memo to have the component pure

File: src/incubator/WheelPicker/index.tsx
Comment (mendyEdri) [reply src/incubator/WheelPicker/index.tsx:14]: Right, auto-import

File: src/incubator/WheelPicker/index.tsx
Comment (mendyEdri) [reply src/incubator/WheelPicker/index.tsx:42]: Do you mean to `onChange: PropTypes.func` without specified values / index?

File: src/incubator/WheelPicker/index.tsx
Comment (mendyEdri) [reply src/incubator/WheelPicker/index.tsx:81]: Not the initial, the current index (`onChange` updates the presenter with the current offset, and the presenter calculates it based on the `selectedValue` )

File: src/incubator/WheelPicker/index.tsx
Comment (mendyEdri) [reply src/incubator/WheelPicker/index.tsx:99]: So it works, it based on the value given as `items` or `children`, I fixed the demo so it be more clear it's not an index..
`_.map(months, value => ({label: value, value: value}));`

File: src/incubator/WheelPicker/index.tsx
Comment (mendyEdri) [reply src/incubator/WheelPicker/index.tsx:148]: I think Android could slow down UI thread sometime, while render fade / gradient views (also in native), so I thought to memo it even though it's not logically necessary.
but I removed from `separators` since it's just a simple view.

WDYT?

**Comment by mendyEdri** [comment]
# Please DO NOT merge this until engine's reanimated release
Depend on engine PR [https://github.com/wix-private/wix-one-app-engine/pull/2123](url)

**Comment by ethanshar** [comment]
@mendyEdri 
The reason this is pending is because we wanted to check it on a real one app version, right? 
Did we tested on one app after the onChange fixes?

**Review by ethanshar — APPROVED** [review]
Approving, don't merge yet. 
Let's figure how we want to rollout this feature carefully.

**Comment by mendyEdri** [comment]
> > @mendyEdri
> > The reason this is pending is because we wanted to check it on a real one app version, right?
> > Did we tested on one app after the onChange fixes?
> 
> @ethanshar, exactly. i've added in the jira-ticket the relevant one-app version with this fix :)


> Approving, don't merge yet.
> Let's figure how we want to rollout this feature carefully.

@ethanshar, We have UX approval for including it without haptic (yet). 
Can I merge?

**Pull request merged**
